backend/src/s9pk/merkle_archive/test.rs

```diff
@@ -0,0 +1,126 @@
+use crate::s9pk::merkle_archive::sink::TrackingWriter;
+use crate::s9pk::merkle_archive::{Entry, EntryContents, MerkleArchive};
+
+#[instrument]
+fn test(files: Vec<(PathBuf, String)>) -> Result<(), Error> {
+    let mut root = DirectoryContents::<Arc<[u8]>>::new();
+    let mut check_set = BTreeMap::<PathBuf, String>::new();
+    for (path, content) in files {
+        if let Err(e) = root.insert_path(
+            &path,
+            Entry::new(EntryContents::File(FileContents::new(
+                content.clone().into_bytes().into(),
+            ))),
+        ) {
+            eprintln!("failed to insert file at {path:?}: {e}");
+        } else {
+            let mut remaining = check_set.split_off(&path);
+            while {
+                if let Some((p, s)) = remaining.pop_first() {
+                    if !p.starts_with(&path) {
+                        remaining.insert(p, s);
+                        false
+                    } else {
+                        true
+                    }
+                } else {
+                    false
+                }
+            } {}
+            check_set.append(&mut remaining);
+            check_set.insert(path.clone(), content);
+        }
+    }
+    let key = SigningKey::generate(&mut rand::thread_rng());
+    let mut archive = MerkleArchive::new(root, key);
+    tokio::runtime::Builder::new_current_thread()
+        .enable_io()
+        .build()
+        .unwrap()
+        .block_on(async move {
+            archive.update_hashes(true).await?;
+            let mut s1 = Vec::new();
+            archive
+                .serialize(&mut TrackingWriter::new(0, &mut s1), true)
+                .await?;
+            let s1: Arc<[u8]> = s1.into();
+            let archive = MerkleArchive::deserialize(&s1, &mut Cursor::new(s1.clone())).await?;
+
+            for (path, content) in check_set {
+                match archive.contents.get_path(&path).map(|e| e.as_contents()) {
+                    Some(EntryContents::File(f)) => {
+                        ensure_code!(
+                            &f.to_vec().await? == content.as_bytes(),
+                            ErrorKind::ParseS9pk,
+                            "File at {path:?} does not match input"
+                        )
+                    }
+                    _ => {
+                        return Err(Error::new(
+                            eyre!("expected file at {path:?}"),
+                            ErrorKind::ParseS9pk,
+                        ))
+                    }
+                }
+            }
+
+            let mut s2 = Vec::new();
+            archive
+                .serialize(&mut TrackingWriter::new(0, &mut s2), true)
+                .await?;
+            let s2: Arc<[u8]> = s2.into();
+            ensure_code!(s1 == s2, ErrorKind::Pack, "s1 does not match s2");
+
+            Ok(())
+        })
+}
+
+proptest::proptest! {
+    #[test]
+    fn property_test(files: Vec<(PathBuf, String)>) {
```

Comment:
rename this function to describe what the property is actually testing. I think it's saying that you're asserting no error is ever thrown from this function? What's curious about that is that you should be able to assert that by dropping the Error from the type signature, no?